File: javascript/leetcode/347. Top K Frequent Elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  const seen = new Map();
  const freq = [];

  nums.forEach((el) => {
    if (seen.has(el)) seen.set(el, seen.get(el) + 1);
    else seen.set(el, 1);
  });

  Array.from(seen.entries()).forEach((el) => {
    freq.push({ num: el[0], freq: el[1] });
  });

  const ans = [];
  freq.sort((a, b) => b.freq - a.freq);

  for (let i = 0; i < k; ++i) ans.push(freq[i].num);

  return ans;
};

File: c++/leetcode/347. Top K Frequent Elements.cpp
class Solution
{
  struct Compare
  {
    bool operator()(pair<int, int> a, pair<int, int> b)
    {
      return a.second < b.second;
    };
  };

public:
  vector<int> topKFrequent(vector<int> &nums, int k)
  {
    unordered_map<int, int> seen;
    priority_queue<pair<int, int>, vector<pair<int, int>>, Compare> freq;

    for (auto i : nums)
      ++seen[i];

    for (auto i : seen)
    {
      freq.push(i);
    }

    vector<int> ans;

    for (int i = 0; i < k; ++i)
    {
      ans.push_back(freq.top().first);
      freq.pop();
    }

    return ans;
  }
};